File: HW/HW11/solution.py
class Solution:
    def floodFill(self, image, sr, sc, newColor):
        if image[sr][sc] == newColor:
            return image
        
        self.flood(image, image[sr][sc], newColor, sr, sc)
        
        return image

    def flood(self, image, oldColor, newColor, sr, sc):
        if (sr < 0 or
            sc < 0 or
            sr >= len(image) or
            sc >= len(image[sr]) or
            image[sr][sc] == newColor or
            image[sr][sc] != oldColor):
            return
        
        image[sr][sc] = newColor
        self.flood(image, oldColor, newColor, sr - 1, sc)
        self.flood(image, oldColor, newColor, sr, sc - 1)
        self.flood(image, oldColor, newColor, sr, sc + 1)
        self.flood(image, oldColor, newColor, sr + 1, sc)

File: HW/HW11/Solution.java
class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        if (image[sr][sc] == newColor) return image;
        
        flood(image, image[sr][sc], newColor, sr, sc);
        
        return image;
    }

   public void floodFill(int[][] image, int sr, int sc, int color) {
    if (
        sr < 0 ||
        sc < 0 ||
        sr >= image.size() ||
        sc >= image[sr].size() ||
        image[sr][sc] == newColor ||
        image[sr][sc] != oldColor
    ) 
        return;
    

    image[sr][sc] = newColor;
    floodFill(image, sr - 1, sc, newColor, oldColor);
    floodFill(image, sr, sc - 1, newColor, oldColor);
    floodFill(image, sr, sc + 1, newColor, oldColor);
    floodFill(image, sr + 1, sc, newColor, oldColor);
}
}
